p11-kit/rpc-client.c

```diff
@@ -406,9 +406,15 @@ mechanism_list_purge (CK_MECHANISM_TYPE_PTR mechs,
 	for (i = 0; i < (int)(*n_mechs); ++i) {
```

Comment:
I might rewrite this into two loops, something like:
```c
/* Trim unsupported mechanisms at the end */
for (; *n_mechs > 0 && !p11_rpc_mechanism_is_supported (mechs[*n_mechs - 1]); --*n_mechs)
  ;

/* Remove unsupported mechanisms in the middle of the array */
for (i = 0; i < *n_mechs; i++) {
  ...
}
```
That way, we don't need a special case.

Reply:
How about going through the list from the back and swapping all unsupported mechs to the end, then truncating n_mechs once? That would give us a lot less copying and move from O(n^2) to O(n) (not that it matters for the ~70 results we're typically getting.

Reply:
Wouldn't that algorithm change the order of elements, though it might be acceptable in this specific case?

Reply:
Yes and yes.

Reply:
Would you like to make the change? Running `git grep memmove` shows some other places that could also benefit from it.

Reply:
On the other hand, `GetMechanismList` may return mechanisms in the most preferred order, allowing the application to optimize the later search. So I would rather keep the original order. I have created a new PR with a slight modification in #456.